man/vattribute.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.41.2.
.TH VATTRIBUTE "8" "May 2013" "vattribute  -- sets/gets attributes of vservers" "System Administration"
.SH NAME
vattribute \- sets/gets attributes of vservers
.SH DESCRIPTION
.SS "Usage:"
.IP
vattribute [\-\-xid <xid>] {\-\-get|\-\-set [\-\-bcap [~!]<cap>] [\-\-ccap [~!]<cap>]
[\-\-flag [~!]<flag>] [\-\-secure]} \fB\-\-\fR [<program> <args>*]
.TP
\fB\-\-bcap\fR <cap>
\&...  system  capability to be set
.TP
\fB\-\-ccap\fR <cap>
\&...  context capability to be set
.TP
\fB\-\-flag\fR <flag>
\&...  context flag to be set
.TP
\fB\-\-umask\fR <mask> ...
unshare mask to be set
.PP
Please report bugs to vserver@list.linux\-vserver.org
.PP
This program is part of util\-vserver
.SH COPYRIGHT
Copyright \(co 2004 Enrico Scholz
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.


NEWS
version 0.30.215
================

	- added support for using libnss instead of beecrypt for
	vhashify's hash calculation.  libnss has bad SHA1 performance on
	i386, but is a) maintained and b) gives better performance with
	all other hashes and on x86_64.

	- added tunctl, a small program to setup tun/tap interfaces, with
	support in the configuration.

	- vattribute and nattribute now support a --get operation, which
	parses the flags and capabilities into a human-readable form.

	- vspace supports all the spaces currently available.

	- vmemctrl controls the OOM bias of context.

	- dynamic contexts are unbroken.

	- hopefully all rpm-fake races have been fixed.

	- flock has been replaced by lockf, which works on NFS.

	- the clone build method now supports exclude lists.

	- vmount handles mounting into a guest.

	- various fixes for the Gentoo-scripts, thanks to Benedikt Böhm.

	- support for yum's persistdir has been added.

	- per-guest device maps are now supported.

	- vserver ... build now sets the barrier on the parent directory.

	- fsck is now supported for the guest's fstab.

	- symlinks are now handled on all levels in namespace cleanup.

	- the debootstrap build method now supports debootstrap 1.0.8.

	- a documentation update, thanks to Micah Anderson.

	- use numeric uids/gids for the template build method.

	- add support for the alpine and f8 distributions.
	
version 0.30.214
================
	- patches for yum-3.2.

	- various fixes to allow for *SuSE guests.

	- added Fedora 7.

	- replace file with h2ext for build -m template.

	- add exclude list support for vclone.

	- include debootstrap scripts for a lot of distributions.

	- fixes to allow the Gentoo initscripts to be used with baselayout 2.

	- remove disk limit caches on vserver ... delete.

	- added support for the new APIs in 2.2 and 2.3, in particular the
	netv2 API, allowing native IPv6 support.

	- added support for a separate filesystem tag.

version 0.30.213
================
	- added support for the device mapping feature present in development
	kernels.

	- further improved Gentoo support, again thanks to Benedikt Böhm.

	- support persistent contexts.

	- add another initscript that creates directories, configures vshelper
	as well as kills all contexts.

	- new vserver-stat implementation, which uses the accounting APIs
	available in Linux-VServer 2.2.

	- added support for setting sysctl values in the guest configuration.

	- added a clone build method, which simply links already unified files.

	- implement dynamic contexts in userspace, generating a static one
	during vserver ... build.

	- fix the endless loop in vlogin.

	- fix Ubuntu builds.

	- improved support for Debian Etch, thanks to Benedikt Böhm.

	- added CentOS 5.

	- use the state change helper to try to cleanup after us if the guest
	exited on its own.

version 0.30.212
================
	- support for the Linux-VServer 2.1 API has been implemented.

	- the scheduler configuration has been split up into multiple files.

	- namespace cleanup is enabled by default. This can be disabled by
	touch ${sysconfdir}/vservers/.defaults/nonamespace-cleanup

	- all unset ulimits are now reset to infinity, so limits are no longer
	inherited from the host.

	- much improved Gentoo support, thanks to Benedikt Böhm.

version 0.30.211
================

	- novlandev is now the default. To reenable the old behaviour of
	creating VLAN interfaces where used, touch
	${sysconfdir}/vservers/.defaults/interfaces/vlandev.

	- vshelper now works correctly with the kernel legacy support turned
	off.

	- vlogin provides a terminal proxy which allocates a new psuedo-tty
	on the inside when you enter a guest.

	- and lots of other features/fixes, see ChangeLog.


version 0.30.210
================

	- LEGACY BREAKAGE: by default, only recent kernel APIs will be
	supported so you have to make sure to use at least the vs2.0.1
	kernel patch. With older kernels, you have to call 'configure'
	with '--enable-apis=NOLEGACY'. Some programs will not work anymore
	with the default settings; these ones are:

	     + capchroot
	     * chcontext-context
	     * reducecap

	- dlimit support was integrated into the standard vserver
	operations. Thanks to Daniel Hokka Zakrisson for his patch.

	- recent kernel API is now used for setting up the network
	contexts. So it should be now possible to use a kernel compiled
	without CONFIG_VSERVER_LEGACYNET. Thanks to Daniel Hokka Zakrisson
	for his patch.  .


version 0.30.209
================

	- large revamping...
	
	
version 0.23.6
==============

	- fixed '--level' option on 'vserver ... chkconfig'


src/vattribute.c
// $Id$    --*- c -*--

// Copyright (C) 2004 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "util.h"
#include "attribute-util.h"
#include <lib/vserver.h>

#include <getopt.h>
#include <stdint.h>
#include <errno.h>

#define ENSC_WRAPPERS_PREFIX	"vattribute: "
#define ENSC_WRAPPERS_VSERVER	1
#define ENSC_WRAPPERS_UNISTD	1
#include <wrappers.h>

#define CMD_HELP		0x1000
#define CMD_VERSION		0x1001
#define CMD_XID			0x2000
#define CMD_SET			0x2001
#define CMD_CCAP		0x2002
#define CMD_FLAG		0x2003
#define CMD_SECURE		0x2004
#define CMD_BCAP		0x2005
#define CMD_GET			0x2006
#define CMD_UMASK		0x2007

int			wrapper_exit_code = 1;

struct option const
CMDLINE_OPTIONS[] = {
  { "help",       no_argument,       0, CMD_HELP },
  { "version",    no_argument,       0, CMD_VERSION },
  { "xid",        required_argument, 0, CMD_XID },
  { "set",        no_argument,       0, CMD_SET },
  { "get",        no_argument,       0, CMD_GET },
  { "ccap",       required_argument, 0, CMD_CCAP },
  { "bcap",       required_argument, 0, CMD_BCAP },
  { "flag",       required_argument, 0, CMD_FLAG },
  { "secure",     no_argument,       0, CMD_SECURE },
  { "umask",      required_argument, 0, CMD_UMASK },
  {0,0,0,0}
};

struct Arguments {
    xid_t		xid;
    struct vc_ctx_flags flags;
    struct vc_ctx_caps  caps;
    struct vc_umask     umask;
    int			mode;
};

static void
showHelp(int fd, char const *cmd, int res)
{
  WRITE_MSG(fd, "Usage:\n    ");
  WRITE_STR(fd, cmd);
  WRITE_MSG(fd,
	    " [--xid <xid>] {--get|--set [--bcap [~!]<cap>] [--ccap [~!]<cap>]\n"
	    "    [--flag [~!]<flag>] [--secure]} -- [<program> <args>*]\n"
	    "\n"
	    " --bcap <cap>   ...  system  capability to be set\n"
	    " --ccap <cap>   ...  context capability to be set\n"
	    " --flag <flag>  ...  context flag to be set\n"
	    " --umask <mask> ...  unshare mask to be set\n"
	    "\n"
	    "Please report bugs to " PACKAGE_BUGREPORT "\n");

  exit(res);
}

static void
showVersion()
{
  WRITE_MSG(1,
	    "vattribute " VERSION " -- sets/gets attributes of vservers\n"
	    "This program is part of " PACKAGE_STRING "\n\n"
	    "Copyright (C) 2004 Enrico Scholz\n"
	    VERSION_COPYRIGHT_DISCLAIMER);
  exit(0);
}

static void
parseFlags(char const *str, struct vc_ctx_flags *flags)
{
  struct vc_err_listparser	err;
  int				rc;

  rc = vc_list2cflag(str,0, &err, flags);
  
  if (rc==-1) {
    WRITE_MSG(2, "Unknown flag '");
    Vwrite(2, err.ptr, err.len);
    WRITE_MSG(2, "'\n");
    exit(wrapper_exit_code);
  }
}

static void
parseBCaps(char const *str, struct vc_ctx_caps *caps)
{
  struct vc_err_listparser	err;
  int				rc;

  rc = vc_list2bcap(str,0, &err, caps);
  
  if (rc==-1) {
    WRITE_MSG(2, "Unknown bcap '");
    Vwrite(2, err.ptr, err.len);
    WRITE_MSG(2, "'\n");
    exit(wrapper_exit_code);
  }
}

static void
parseCCaps(char const *str, struct vc_ctx_caps *caps)
{
  struct vc_err_listparser	err;
  int				rc;

  rc = vc_list2ccap(str,0, &err, caps);
  
  if (rc==-1) {
    WRITE_MSG(2, "Unknown ccap '");
    Vwrite(2, err.ptr, err.len);
    WRITE_MSG(2, "'\n");
    exit(wrapper_exit_code);
  }
}

static void
parseUMask(char const *str, struct vc_umask *umask)
{
  struct vc_err_listparser	err;
  int				rc;

  rc = vc_list2umask(str, 0, &err, umask);
  
  if (rc==-1) {
    WRITE_MSG(2, "Unknown namespace '");
    Vwrite(2, err.ptr, err.len);
    WRITE_MSG(2, "'\n");
    exit(wrapper_exit_code);
  }
}

static void
parseSecure(struct vc_ctx_flags UNUSED * flags,
	    struct vc_ctx_caps  UNUSED * caps)
{
  caps->ccaps = ~vc_get_insecureccaps();
  caps->cmask = ~0ull;
  caps->bcaps = ~vc_get_insecurebcaps();
  caps->bmask = ~0ull;

    // TODO: generalize this
  flags->flagword = VC_VXF_HIDE_NETIF;
  flags->mask     = VC_VXF_HIDE_NETIF;
}

static int
printAttrs(struct Arguments *args)
{
  struct vc_ctx_flags flags;
  struct vc_ctx_caps caps;
  struct vc_umask umask = { .mask = ~0, .umask = 0x20200 };

  Evc_get_cflags(args->xid, &flags);
  Evc_get_ccaps(args->xid, &caps);

  print_bitfield(1, bcap, "bcapabilities", &caps.bcaps);
  print_bitfield(1, ccap, "ccapabilities", &caps.ccaps);
  print_bitfield(1, cflag, "flags", &flags.flagword);
  if (vc_get_umask(args->xid, &umask) == 0) {
    print_bitfield(1, umask, "umask", &umask.umask);
  }

  return 0;
}

int main(int argc, char *argv[])
{
  struct Arguments		args = {
    .xid   = VC_NOCTX,
    .flags = { .flagword = 0, .mask = 0 },
    .caps  = { .bcaps = 0, .bmask = 0,.ccaps = 0, .cmask = 0 },
    .umask = { .umask = 0, .mask = 0 },
    .mode  = CMD_SET,
  };

  while (1) {
    int		c = getopt_long(argc, argv, "+", CMDLINE_OPTIONS, 0);
    if (c==-1) break;

    switch (c) {
      case CMD_HELP	:  showHelp(1, argv[0], 0);
      case CMD_VERSION	:  showVersion();
      case CMD_SET	:  args.mode = CMD_SET;                    break;
      case CMD_GET	:  args.mode = CMD_GET;                    break;
      case CMD_XID	:  args.xid = Evc_xidopt2xid(optarg,true); break;
      case CMD_FLAG	:  parseFlags(optarg, &args.flags);        break;
      case CMD_CCAP	:  parseCCaps(optarg, &args.caps);         break;
      case CMD_BCAP	:  parseBCaps(optarg, &args.caps);         break;
      case CMD_SECURE	:  parseSecure(&args.flags, &args.caps);   break;
      case CMD_UMASK	:  parseUMask(optarg, &args.umask);        break;
      default		:
	WRITE_MSG(2, "Try '");
	WRITE_STR(2, argv[0]);
	WRITE_MSG(2, " --help' for more information.\n");
	return 255;
	break;
    }
  }

  if (args.xid==VC_NOCTX) args.xid = Evc_get_task_xid(0);

  if (args.mode == CMD_SET) {
    if ((args.caps.cmask || args.caps.bmask) &&
	vc_set_ccaps(args.xid, &args.caps)==-1)
      perror(ENSC_WRAPPERS_PREFIX "vc_set_ccaps()");
    else if (args.flags.mask &&
	     vc_set_cflags(args.xid, &args.flags)==-1)
      perror(ENSC_WRAPPERS_PREFIX "vc_set_flags()");
    else if (args.umask.mask &&
             vc_set_umask(args.xid, &args.umask)==-1)
      perror(ENSC_WRAPPERS_PREFIX "vc_set_umask()");
    else if (optind<argc)
      EexecvpD(argv[optind], argv+optind);
    else
      return EXIT_SUCCESS;
  }
  else if (args.mode == CMD_GET) {
    printAttrs(&args);
    if (optind<argc)
      EexecvpD(argv[optind], argv+optind);
    else
      return EXIT_SUCCESS;
  }

  return EXIT_FAILURE;
}
